helpers/preview/bin/preview-rst

```diff
@@ -0,0 +1,313 @@
+    # that refreshing the browser works to pick up changes if the file is
+    # modified.
+    input_dirname=`dirname "${input_file}"`
+    output_dir_subpath=`realpath "${input_dirname}"`
+    # Namespace our temporary directories
+    output_dir="/tmp/${script_basename}/output${output_dir_subpath}"
+    # This path should remain constant for subsequent HTML generations of the
+    # same source file path (see above). We leave deletion of this temporary
+    # directory up to the user or the operating system.
+    mkdir -p "${output_dir}"
+    echo "${output_dir}"
+}
+
+function progs_exist {
+    progs="${1}"
+    return_code=0
+    for prog in "${progs}"; do
+        if test -z `which "${prog}"`; then
+            return_code=1
+            echo "The `${prog}` program is not on your PATH." >&2
+        fi
+    done
+    return $return_code
+}
+
+
+# This function creates a temporary Python virtual environment that is shared
+# by every invokation of this script
+function get_venv_dir {
+    tmp_dir="${1}"
+    # Namespace our temporary directories
+    venv_dir="/tmp/${script_basename}/deps/${PY}/venv"
+    log_dir="/tmp/${script_basename}/deps/${PY}/log"
+    log_file="${log_dir}/pip.log"
+    if test ! -f "${venv_dir}/bin/activate"; then
+        mkdir -p "${venv_dir}" "${log_dir}"
+        ${PY} -m venv "${venv_dir}" \
+            >> "${log_file}"
+        . "${venv_dir}/bin/activate"
+        # The only way for a user to get dependency upgrades is to remove the
+        # `venv` directory.
+        #
+        # TODO: Add some method of handling upgrades that doesn't involve
+        # manual intervention by the user. For example, adding a stamp file and
+        # using modified date to perform upgrades once a month. The main issue
+        # here is that upgrading dependencies takes time, and would add latency
+        # to the preview script. Another option would be forking to perform the
+        # upgrades.
+        pip install --upgrade pip \
+            >> "${log_file}" || rm -r "${venv_dir}"
+        pip install --upgrade -r "${etc_dir}/requirements.txt" \
+            >> "${log_file}" || rm -r "${venv_dir}"
+    fi
+    if test ! -f "${venv_dir}/bin/activate"; then
+        # The pip installation failed for some reason
+        exit 1
+    fi
+    echo "${venv_dir}"
+}
+
+
+# Display strategies
+# =============================================================================
+
+# This function assumes you already have `sphinx-autobuild` running, and opens
+# the corresponding HTML file in a browser.
+#
+# TODO: In a later update, we might want to check whether this is the case, and
+# if not, start one up. This should respect the Makefile, if one exists, or
+# else invoke `sphinx-autobuild` manually.
+function display_sphinx_autobuild {
+    rst_file="${1}"
+    html_file=`echo "${rst_file}" | \
+        sed 's,.*/docs/,http://127\.0\.0\.1:8000/,' | \
```

Comment:
Here, port 8000 is referenced. However, `sphinx-autobuild` is launched on port 8080. Is this intended?

Reply:
this function is specifically about opening a browser to an already-running sphinx-autobuild that you started by hand (e.g., `make dev`) which runs on port 8000. this makes it super easy to just run `make dev` and then in any file hit your "preview" shortcut and go directly to that file in your docs project